File: shell/gnome/src/prefs/registry.js
import { _, fill } from '../i18n.js';

const SCHEMA_VERSION = 1;
export const LABEL_MAX_CHARS = 64;
const PROVIDER_ID = /^[a-z0-9_-]+$/;

export class RegistryError extends Error {}

function isObject(value) {
    return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function text(value) {
    return typeof value === 'string' && value.trim().length > 0 ? value.trim() : null;
}

function httpsUrl(value) {
    const url = text(value);
    return url && /^https:\/\/\S+$/.test(url) ? url : null;
}

function parseMethod(raw) {
    if (!isObject(raw)) return null;
    if (raw.kind === 'cli_login') return { kind: 'cli_login', program: text(raw.program) };
    if (raw.kind === 'api_key')
        return { kind: 'api_key', label: text(raw.label), consoleUrl: httpsUrl(raw.console_url), hint: text(raw.hint) };
    if (raw.kind === 'auto_detect') return { kind: 'auto_detect', reason: text(raw.reason) };
    return null;
}

function parseLinks(raw) {
    const links = isObject(raw) ? raw : {};
    return { status: httpsUrl(links.status), dashboard: httpsUrl(links.dashboard), usage: httpsUrl(links.usage) };
}

function parseProvider(raw) {
    if (!isObject(raw) || typeof raw.id !== 'string' || !PROVIDER_ID.test(raw.id)) return null;
    const methods = (Array.isArray(raw.add_account) ? raw.add_account : []).map(parseMethod).filter(Boolean);
    if (methods.length === 0) return null;
    return {
        id: raw.id,
        displayName: text(raw.display_name) ?? raw.id,
        methods,
        multiAccount: raw.multi_account === true,
        links: parseLinks(raw.links),
    };
}

function decode(json) {
    try {
        return JSON.parse(json);
    } catch (error) {
        throw new RegistryError(
            fill(_('Unreadable provider list from the Headroom service: {reason}'), { reason: error.message })
        );
    }
}

export function parseProviders(json) {
    const raw = decode(json);
    if (!isObject(raw)) throw new RegistryError(_('Unexpected provider list from the Headroom service'));
    if (raw.version !== SCHEMA_VERSION)
        throw new RegistryError(
            fill(_('Headroom service lists providers in version {actual}, expected {expected}'), {
                actual: raw.version,
                expected: SCHEMA_VERSION,
            })
        );
    const providers = (Array.isArray(raw.providers) ? raw.providers : []).map(parseProvider).filter(Boolean);
    return providers.filter((provider, index) => providers.findIndex(other => other.id === provider.id) === index);
}

export function methodSummary(method) {
    if (method.kind === 'cli_login')
        return method.program ? fill(_('Sign in with {program}'), { program: method.program }) : _('Sign in');
    if (method.kind === 'api_key') return _('API key');
    return _('Detected automatically');
}

export function addAccountArgs(providerId, method, label) {
    const keyFlag = method.kind === 'api_key' ? ['--api-key-stdin'] : [];
    const trimmed = label.trim();
    return ['accounts', 'add', providerId, ...keyFlag, ...(trimmed ? [`--label=${trimmed}`] : [])];
}

export function loginArgs(accountId, method) {
    return ['accounts', 'login', accountId, ...(method.kind === 'api_key' ? ['--api-key-stdin'] : [])];
}

export function loginMethod(provider) {
    return (
        provider.methods.find(method => method.kind === 'cli_login') ??
        provider.methods.find(method => method.kind === 'api_key') ??
        null
    );
}

export function signInTarget(account, providers) {
    if (account.recovery?.action !== 'sign_in') return null;
    const provider = (providers ?? []).find(entry => entry.id === account.provider);
    if (!provider || !loginMethod(provider)) return null;
    return { provider, loginId: account.recovery.accountId ?? account.id };
}

export function providerSummary(provider) {
    return provider.methods.map(methodSummary).join(' · ');
}

File: crates/headroom-tray/src/preferences/registry.rs
use serde::Deserialize;

use crate::i18n::{Lang, fill};

const SCHEMA_VERSION: u32 = 1;

#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
pub enum RegistryError {
    #[error("unreadable provider list from the Headroom service: {0}")]
    Unreadable(String),
    #[error("the Headroom service lists providers in version {0}, expected {SCHEMA_VERSION}")]
    Version(u32),
}

impl RegistryError {
    #[must_use]
    pub fn message(&self, lang: Lang) -> String {
        match self {
            RegistryError::Unreadable(reason) => fill(
                lang.tr("Unreadable provider list from the Headroom service: {reason}"),
                &[("reason", reason)],
            ),
            RegistryError::Version(actual) => fill(
                lang.tr(
                    "Headroom service lists providers in version {actual}, expected {expected}",
                ),
                &[
                    ("actual", &actual.to_string()),
                    ("expected", &SCHEMA_VERSION.to_string()),
                ],
            ),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AddMethod {
    CliLogin {
        program: Option<String>,
    },
    ApiKey {
        label: Option<String>,
        console_url: Option<String>,
        hint: Option<String>,
    },
    AutoDetect {
        reason: Option<String>,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct ProviderLinks {
    pub status: Option<String>,
    pub dashboard: Option<String>,
    pub usage: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ProviderInfo {
    pub id: String,
    pub display_name: String,
    pub methods: Vec<AddMethod>,
    pub links: ProviderLinks,
}

#[derive(Debug, Deserialize)]
struct RawRegistry {
    version: u32,
    #[serde(default)]
    providers: Vec<serde_json::Value>,
}

#[derive(Debug, Deserialize)]
struct RawProvider {
    id: String,
    #[serde(default)]
    display_name: Option<String>,
    #[serde(default)]
    add_account: Vec<RawMethod>,
    #[serde(default)]
    links: Option<serde_json::Value>,
}

#[derive(Debug, Default, Deserialize)]
struct RawLinks {
    #[serde(default)]
    status: Option<String>,
    #[serde(default)]
    dashboard: Option<String>,
    #[serde(default)]
    usage: Option<String>,
}

#[derive(Debug, Deserialize)]
struct RawMethod {
    kind: String,
    #[serde(default)]
    program: Option<String>,
    #[serde(default)]
    label: Option<String>,
    #[serde(default)]
    console_url: Option<String>,
    #[serde(default)]
    hint: Option<String>,
    #[serde(default)]
    reason: Option<String>,
}

fn text(value: Option<String>) -> Option<String> {
    value
        .map(|text| text.trim().to_owned())
        .filter(|text| !text.is_empty())
}

fn https_url(value: Option<String>) -> Option<String> {
    text(value).filter(|url| url.starts_with("https://") && !url.contains(char::is_whitespace))
}

fn valid_id(id: &str) -> bool {
    !id.is_empty()
        && id
            .chars()
            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_' || c == '-')
}

fn method(raw: RawMethod) -> Option<AddMethod> {
    match raw.kind.as_str() {
        "cli_login" => Some(AddMethod::CliLogin {
            program: text(raw.program),
        }),
        "api_key" => Some(AddMethod::ApiKey {
            label: text(raw.label),
            console_url: https_url(raw.console_url),
            hint: text(raw.hint),
        }),
        "auto_detect" => Some(AddMethod::AutoDetect {
            reason: text(raw.reason),
        }),
        _ => None,
    }
}

fn links(value: Option<serde_json::Value>) -> ProviderLinks {
    let raw: RawLinks = value
        .and_then(|value| serde_json::from_value(value).ok())
        .unwrap_or_default();
    ProviderLinks {
        status: https_url(raw.status),
        dashboard: https_url(raw.dashboard),
        usage: https_url(raw.usage),
    }
}

fn provider(value: serde_json::Value) -> Option<ProviderInfo> {
    let raw: RawProvider = serde_json::from_value(value).ok()?;
    if !valid_id(&raw.id) {
        return None;
    }
    let methods: Vec<AddMethod> = raw.add_account.into_iter().filter_map(method).collect();
    if methods.is_empty() {
        return None;
    }
    Some(ProviderInfo {
        display_name: text(raw.display_name).unwrap_or_else(|| raw.id.clone()),
        id: raw.id,
        methods,
        links: links(raw.links),
    })
}

pub fn parse_providers(json: &str) -> Result<Vec<ProviderInfo>, RegistryError> {
    let raw: RawRegistry =
        serde_json::from_str(json).map_err(|error| RegistryError::Unreadable(error.to_string()))?;
    if raw.version != SCHEMA_VERSION {
        return Err(RegistryError::Version(raw.version));
    }
    let mut providers: Vec<ProviderInfo> = Vec::new();
    for info in raw.providers.into_iter().filter_map(provider) {
        if providers.iter().all(|known| known.id != info.id) {
            providers.push(info);
        }
    }
    Ok(providers)
}

#[must_use]
pub fn method_summary(lang: Lang, method: &AddMethod) -> String {
    match method {
        AddMethod::CliLogin {
            program: Some(program),
        } => fill(lang.tr("Sign in with {program}"), &[("program", program)]),
        AddMethod::CliLogin { program: None } => lang.tr("Sign in").to_owned(),
        AddMethod::ApiKey { .. } => lang.tr("API key").to_owned(),
        AddMethod::AutoDetect { .. } => lang.tr("Detected automatically").to_owned(),
    }
}

#[must_use]
pub fn provider_summary(lang: Lang, provider: &ProviderInfo) -> String {
    provider
        .methods
        .iter()
        .map(|method| method_summary(lang, method))
        .collect::<Vec<_>>()
        .join(" · ")
}

#[must_use]
pub fn add_account_args(provider_id: &str, method: &AddMethod, label: &str) -> Vec<String> {
    let mut args: Vec<String> = vec!["accounts".into(), "add".into(), provider_id.into()];
    if matches!(method, AddMethod::ApiKey { .. }) {
        args.push("--api-key-stdin".into());
    }
    let label = label.trim();
    if !label.is_empty() {
        args.push(format!("--label={label}"));
    }
    args.extend(["--progress".into(), "json".into()]);
    args
}

#[cfg(test)]
#[path = "registry_tests.rs"]
mod tests;
